### PR TITLE
[4.0][a11y] Amenu

### PR DESCRIPTION
**Pull request by ylahav**
Pull Request for Issue # .

### Summary of Changes

adding keyboard navigation to menu

### Testing Instructions

create different types of menus (with sub-menus and without)
Reach (with 'tab') the menu under test and check the following behavior:

- right arrow - next item (same level). If last - back to first item
- left arrow - prevision item (same level) . if first - go to last item
- down arrow - if sub-menu exist - go to sub level, first item
                      - if not - behave as right arrow
- up arrow - if first item in sub-menu - go to parent item
                 - if not first item or no parent item - same as left arrow

### Expected result

The behavior was explained above

### CODE REVIEW

**Comment by brianteeman** [comment]
me thinks there are some files missing

**Comment by ylahav** [comment]
@brianteeman - Yes, I realize it... I will commit the missing file soon

File: build/media/mod_menu/js/menu.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+document.addEventListener('DOMContentLoaded', function() {
+  var navs = document.querySelectorAll('.nav');
+  [].forEach.call(navs, function(nav) {
+    new setup_navigation(nav);
+  });
+});
+
+var setup_navigation = function(nav, settings) {
+  var settings = { menuHoverClass: 'show-menu', dir: 'ltr' };
+	// Add ARIA role to menubar and menu items
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: commented code should not be committed

File: build/media/mod_menu/js/menu.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        ulChild.classList.add(settings.menuHoverClass);
+      }
+    });
+
+    topLevelEl.addEventListener('mouseout', function(event) {
+      var curEl = event.target;
+      var ulChild = curEl.querySelector('ul');
+      if (ulChild) {
+        ulChild.setAttribute('aria-hidden', 'true');
+        ulChild.setAttribute('aria-hidden', 'true');
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: duplicate line

File: build/media/mod_menu/js/menu.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+              if (settings.dir === "rtl") {
+                prevLiEl.children[0].focus();
+              } else {
+                nextLiEl.children[0].focus();
+              }
+            break;
+          case 'ArrowUp':
+              event.preventDefault();
+              var parent = curLiEl.parentElement.parentElement;
+              console.log(parent.nodeName);
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: console log should not be in production code

File: build/media/mod_menu/js/menu.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+              } else {
+                prevLiEl.children[0].focus();
+              }
+              break;
+          case 'ArrowDown':
+            event.preventDefault();
+            if (curLiEl.classList.contains("parent")) {
+              var child = curLiEl.querySelector('ul');
+              if (child != null) {
+                console.log("child: " + child);
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: console log should not be in production code

File: build/media/mod_menu/js/menu.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                var childLi = child.querySelector('li');
+                console.log("childLi: " + childLi);
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: console log should not be in production code

File: build/media/mod_menu/js/menu.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    ulChild.setAttribute('aria-hidden', 'false');
+    ulChild.classList.add(settings.menuHoverClass);
+  }
+}
+
+function topLevelMouseOut (el, settings) {
+  var ulChild = el.querySelector('ul');
+  if (ulChild) {
+    ulChild.setAttribute('aria-hidden', 'true');
+    ulChild.setAttribute('aria-hidden', 'true');
```
Comment (brianteeman) [inline build/media/mod_menu/js/menu.js]: duplicate line

**Comment by dgrammatiko** [comment]
Please move the script to media_src and convert it to ES6

**Comment by brianteeman** [comment]
This code does a lot more than adding the keyboard arrow support you describe. It is also adding various markup that really should be part of the view not the js - otherwise it cant easily be overriden.  Also you really should have documented those markup changes in the original report so that they can be tested correctly.

**Comment by ylahav** [comment]
> 
> 
> This code does a lot more than adding the keyboard arrow support you describe. It is also adding various markup that really should be part of the view not the js - otherwise it cant easily be overriden. Also you really should have documented those markup changes in the original report so that they can be tested correctly.

I Agree. The markup left from a plugin I am writing to old versions of Joomla where markup will be added as well... I will remove it from the code.

**Comment by dgrammatiko** [comment]
FWIW this is still not es6. Try to lint the script and you’ll figure out what needs to be done

**Comment by ylahav** [comment]
> FWIW this is still not es6. Try to lint the script and you’ll figure out what needs to be done

I did it with  - https://jshint.com. can you recommended another lint?

**Comment by dgrammatiko** [comment]
Just exetute the one in the repo:
- `npm run lint:js`

**Comment by wilsonge** [comment]
@dgrammatiko our linter runs fine against this too (see the drone results since i just merged in staging)

EDIT: Ignore me @ylahav can you edit the file name to be `.es6.js` please

**Comment by wilsonge** [comment]
Updated to joomla code standards

**Comment by dgrammatiko** [comment]
This is still not ES6...

**Comment by wilsonge** [comment]
Then you need to explain in more detail what the issue is because as you can see from the drone output, it is happy with the linter

**Comment by ylahav** [comment]
I agree with Brian and will fix it. No need to explain...

**Comment by brianteeman** [comment]
Surely the js should be included in the tmpl - otherwise it can't be overridden

See this as an example https://github.com/joomla/joomla-cms/blob/0d68d9e23fcb7ba17048cc20b2a2b1c56afc1c46/modules/mod_finder/tmpl/default.php#L47

File: modules/mod_menu/mod_menu.php [hunk 7]
```diff
@@ -9,6 +9,7 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\HTML\HTMLHelper;
```
Comment (infograf768) [inline modules/mod_menu/mod_menu.php]: Now useless

**Comment by ylahav** [comment]
What more I have to do in order this change will be included in J4?

**Comment by brianteeman** [comment]
It needs tests - presumably as the a11y team leader you can get the team members to test this

**Comment by dgrammatiko** [comment]
@ylahav FWIW having 3 different scripts to render a menu is not the right way...

**Comment by ylahav** [comment]
> @ylahav FWIW having 3 different scripts to render a menu is not the right way...

@dgrammatiko I know only about one - menu.es6.js (that after George rename mine)... what I miss here, thanks!

**Comment by ylahav** [comment]
> It needs tests - presumably as the a11y team leader you can get the team members to test this

@brianteeman - thanks - I asked them

**Comment by dgrammatiko** [comment]
Yes ther is one more in the module menu and another script called metis menu. Also I’m not sure if this menu doesn’t have any dependency on bootstrap.js as I’m observing some classes there.
And finally some comments on the actual code: you’re traversing the dom way too much you shouldn’t! This will be laggy in any low end mobile

**Comment by brianteeman** [comment]
Should it even be active on a mobile?

**Comment by ylahav** [comment]
> Should it even be active on a mobile?

Good question - I don't think it "give" something if it is active in mobile....

**Comment by ylahav** [comment]
@dgrammatiko - metismenu is used only in the administrator menu... so for the FE it is not applicable...

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on c56c089542743a7fe64b9ece709ba6cc7da06d4a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23254">issues.joomla.org/tracker/joomla-cms/23254</a>.</sub>

File: build/media_src/mod_menu/js/menu.es6.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+          break;
+        default:
+          break;
+      }
+    });
+  });
+}
+
+document.addEventListener('DOMContentLoaded', () => {
+  const navs = document.querySelectorAll('.nav');
```
Comment (bembelimen) [inline build/media_src/mod_menu/js/menu.es6.js]: I'm not sure here, if the check for ```.nav```is a bit too generic? What happens if someone has somewhere a module which uses this class?

Comment (brianteeman) [reply build/media_src/mod_menu/js/menu.es6.js]: Good spot - also a problem if the menu uses a different class

I would suggest that it instead looks for the <nav> element which is what a menu should be wrapped in. Unfortunately the current site menu module does not - it should

**Comment by Hackwar** [comment]
I have tested this item :white_check_mark: successfully on c56c089542743a7fe64b9ece709ba6cc7da06d4a<br /><br />I can tab through the page and navigate through the menu. Looks nice.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23254">issues.joomla.org/tracker/joomla-cms/23254</a>.</sub>

File: build/media_src/mod_menu/js/menu.es6.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
+/**
+ * @copyright  Copyright (C) 2005 - 2018 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
```
Comment (Fedik) [inline build/media_src/mod_menu/js/menu.es6.js:5]: please wrap your code in to anonymous function:
```javascript
(() => {
  'use strict';

  .. your code here

})();
```
thanks

**Comment by fancyFranci** [comment]
@ylahav I would really like to see this merged. Please could you solve the remaining issues? :)

**Comment by ylahav** [comment]
I am in vacation until March 10... Will do it when coming back home..

**Comment by bembelimen** [comment]
@ylahav can we support you in any way? I really like what you did here and would like to have it in Joomla!

**Comment by HLeithner** [comment]
@ylahav are you back and can fix the remaining issues?

**Comment by ylahav** [comment]
> @ylahav are you back and can fix the remaining issues?

Yes, I am back and will fixed it this week

**Comment by wilsonge** [comment]
I'm merging this as it is to clear the path for those who want this in. If someone could make a PR here to solve the issue mentioned here with the target element it would be useful https://github.com/joomla/joomla-cms/pull/23254#discussion_r248213104 

Thanks @ylahav !

**Comment by brianteeman** [comment]
grrh - what was the rush to merge broken code

**Comment by wilsonge** [comment]
What's broken about it? I see the anonymous function issue addressed since the 2 good tests. The fact it's got a slightly generic selector didn't seem to cause any issues when testing - it's sub-optimal for sure - but doesn't appear to be breaking anything immediately. And it's simple to add a fix for.

**Comment by wilsonge** [comment]
I retract the above. After merging in 4.0 the file was in the wrong place - fixed - https://github.com/joomla/joomla-cms/commit/d2098c04b6a9f85510c90840b0c4de244e613c52

**Comment by brianteeman** [comment]
I still dont see what the rush was

**Comment by dgrammatiko** [comment]
Still broken for IE. IE doesn't support iteration over node collections (eg document.querySelectAll IS NOT an array and you cannot iterate it on IE).

Also, this script is traversing the DOM on each event, this is plain wrong as is a performance hog if you ever get a reasonable menu...

Also not IIFE as requested by @Fedik...

**Comment by brianteeman** [comment]
As it has only just been merged please revert it so that it can be fixed BEFORE it enters the codebase

**Comment by mbabker** [comment]
> Also not IIFE as requested by @Fedik...

Yes it is.  The change request was not marked reviewed, completed, or whatnot.

> Still broken for IE. IE doesn't support iteration over node collections (eg document.querySelectAll IS NOT an array and you cannot iterate it on IE).

I'm not seeing this issue noted in this PR, so kind of hard for someone to address an unknown issue.  That would've been helpful feedback in December with your "this is not ES6" comments.

**Comment by brianteeman** [comment]
@wilsonge you have committed this with the git commit message of
[4.0][a11y] Improve accessibility of the admin menu (#23254)

This is NOT for the admin!!

**Comment by brianteeman** [comment]
I can confirm that this does not work in 
IE11
Edge

Console with Edge reports SCRIPT5022: SCRIPT5022: SyntaxError

**Comment by bembelimen** [comment]
I looked into the JS and fixed most of the IE11 issues:
- querySelectorAll => [].slice
- correct selector (instead of .nav)

Still working on:
- :scope call, which does not work in IE11
- the classes where I'm not sure if they're needed

Will do a PR when everything is resolved

**Comment by brianteeman** [comment]
thank you

**Comment by dgrammatiko** [comment]
@bembelimen fwiw I would create a class and in the constructor, I would cache ALL the elements in some arrays. Then the event listeners won't have to traverse the DOM on each call, instead, they will do whatever is needed to the appropriate element(s) from the cache. Properly done...

**Pull request merged**